File: HashMap/454. 4Sum II/Solution.java
class Solution {
    //1. use map to store count of sum for A and B
    //2. for for C and D to find match sum
    public int fourSumCount(int[] A, int[] B, int[] C, int[] D) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int a:A){
            for (int b:B){
                int sum=a+b;
                map.put(sum, map.getOrDefault(sum, 0)+1);
            }
        }
        int count=0;
        for (int c:C){
            for (int d:D){
                int sum=c+d;
                count += map.getOrDefault(-sum, 0);
            }
        }
        return count;
    }
}

File: HashMap/454. 4Sum II/Solution.py
class Solution(object):
    def fourSumCount(self, A, B, C, D):
        """
        :type A: List[int]
        :type B: List[int]
        :type C: List[int]
        :type D: List[int]
        :rtype: int
        """
        prefix = dict()
        for a in A:
            for b in B:
                res = a+b
                prefix[res] = prefix.get(res,0)+1
        count=0
        for c in C:
            for d in D:
                res=c+d
                count += prefix.get(-res,0)
        return count
        